src/crypto/public_key/dsa.js

```diff
@@ -18,111 +18,106 @@
 // A Digital signature algorithm implementation
 
 /**
+ * @requires bn.js
  * @requires crypto/hash
- * @requires crypto/public_key/jsbn
  * @requires crypto/random
+ * @requires config
  * @requires util
  * @module crypto/public_key/dsa
  */
 
-import BigInteger from './jsbn.js';
-import random from '../random.js';
-import hashModule from '../hash';
-import util from '../../util.js';
+import BN from 'bn.js';
+import hash from '../hash';
+import random from '../random';
 import config from '../../config';
+import util from '../../util';
 
-export default function DSA() {
-  // s1 = ((g**s) mod p) mod q
-  // s1 = ((s**-1)*(sha-1(m)+(s1*x) mod q)
-  function sign(hashalgo, m, g, p, q, x) {
+const two = new BN(2);
+const zero = new BN(0);
+
+/*
+  TODO regarding the hash function, read:
+   https://tools.ietf.org/html/rfc4880#section-13.6
+   https://tools.ietf.org/html/rfc4880#section-14
+*/
+
+export default {
+  /*
+   * hash_algo is integer
+   * m is string
+   * g, p, q, x are all BN
+   * returns { r: BN, s: BN }
+   */
+  sign: function(hash_algo, m, g, p, q, x) {
+    let k;
+    let r;
+    let s;
+    let t;
+    const redp = new BN.red(p);
+    const redq = new BN.red(q);
+    const gred = g.toRed(redp);
+    const xred = x.toRed(redq);
     // If the output size of the chosen hash is larger than the number of
     // bits of q, the hash result is truncated to fit by taking the number
     // of leftmost bits equal to the number of bits of q.  This (possibly
     // truncated) hash function result is treated as a number and used
     // directly in the DSA signature algorithm.
-    const hashed_data = util.getLeftNBits(util.Uint8Array2str(hashModule.digest(hashalgo, util.str2Uint8Array(m))), q.bitLength());
-    const hash = new BigInteger(util.hexstrdump(hashed_data), 16);
+    const h = new BN(
+      util.str2Uint8Array(
+        util.getLeftNBits(
+          util.Uint8Array2str(hash.digest(hash_algo, m)), q.bitLength())));
     // FIPS-186-4, section 4.6:
     // The values of r and s shall be checked to determine if r = 0 or s = 0.
     // If either r = 0 or s = 0, a new value of k shall be generated, and the
     // signature shall be recalculated. It is extremely unlikely that r = 0
     // or s = 0 if signatures are generated properly.
-    let k;
-    let s1;
-    let s2;
     while (true) {
-      k = random.getRandomBigIntegerInRange(BigInteger.ONE, q.subtract(BigInteger.ONE));
-      s1 = (g.modPow(k, p)).mod(q);
-      s2 = (k.modInverse(q).multiply(hash.add(x.multiply(s1)))).mod(q);
-      if (s1 !== 0 && s2 !== 0) {
-        break;
+      // TODO confirm this range
```

Comment:
According to this: https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.186-4.pdf - k should be in [1, q-1] inclusive (go to Appendix B).